sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,841 @@
+corresponding SPIR-V functionality, while emulation is supposed to be
+implemented through transforming specialization constants into kernel arguments.
+
+In DPC++ Headers/DPC++ RT we don't know a lot of necessary information about
+specialization constants, like: which numeric ID is used for particular
+specialization constant (since we support `SYCL_EXTERNAL`, those IDs can only
+be allocated by the compiler during link stage) or which kernel argument is used
+to pass particular specialization constant (because they are not explicitly
+captured by SYCL kernel functions and regular mechanism for kernel arguments
+handling can't be used here).
+
+Therefore, we can't have headers-only implementation and the crucial part of
+design is how to organize mapping mechanism between SYCL identifiers for
+specialization constants (`specialization_id`s) and low-level identifiers
+(numeric IDs in SPIR-V or information about corresponding kernel arguments).
+
+That mapping mechanism is particularly tricky, because of some additional
+complexity coming from SYCL 2020 specification:
+- `specialization_id` variables, which are used as specialization constant
+  identifiers (being non-type template parameters of some methods) can't be
+  forward-declared in general case (for example, if defined as `static`), which
+  means that we can't use integration header to attach some information to them
+  through some C++ templates tricks (like it is done for regular kernel
+  arguments or kernel names, for example).
+- they also can be declared as `static` or just non-`inline` `constexpr`, which
+  means that they have internal linkage and can't be referenced from other
+  translation units, which means that we can't for example create a new
+  translation unit which contains some mapping from `specialization_id` address
+  to some desired info.
+
+Based on those limitations, the following mapping design is proposed:
+- DPC++ RT uses special function:
+  ```
+  namespace detail {
+    template<auto &SpecName>
+    const char *get_spec_constant_symbolic_ID();
+  }
+  ```
+  Which is only declared, but not defined in there and used to retrieve required
+  information like numeric ID of a specialization constant.
+- Definition of that function template are provided by DPC++ FE in form of
+  _integration footer_: the compiler generates a piece of C++ code which is
+  injected at the end of a translation unit:
+  ```
+  namespace detail {
+    // assuming user defined and used the following specialization_id:
+    // constexpr specialiation_id<int> int_const;
+    // class Wrapper {
+    // public:
+    //   static constexpr specialization_id<float> float_const;
+    // };
+
+    template<>
+    const char *get_spec_constant_symbolic_ID<int_const>() {
+      return "unique_name_for_int_const";
+    }
+    template<>
+    const char *get_spec_constant_symbolic_ID<Wrapper::float_const>() {
+      return "unique_name_for_Wrapper_float_const";
+    }
```

Comment:
Somewhere (either here are below in "DPC++ Compiler: front-end") there should be a description about the requirements for the strings that are returned by `get_spec_constant_symbolic_ID()`.  I think the requirement is:

* If the `specialization_id` variable has internal linkage, the string must be unique across all translation units.  Since the mangled name is not necessarily unique, you cannot use that.  Instead, the FE probably needs to generate a GUID.

* If the `specialization_id` variable has external linkage, the string must be consistent in all translation units that reference this same variable.  The mangled name would be a good choice because it will have the same value in every translation unit that references this same variable.

Reply:
I've added those requirements in ce4788a2f1c1556042f9cc356a6eea4e1e528aba. Note: I've also added a requirement that the built-in should return the same unique string if it was called twice within a single translation unit for the same variable, i.e. the compiler should preserve a mapping `internal linkage variable` -> `GUID` in order to do that.

This additional requirement is related to another conversation we have [below in this PR](https://github.com/intel/llvm/pull/3331#discussion_r605907197) - let's wait for some feedback from @erichkeane about what what be easier for FE to implement - stable unique names for variable with internal linkage or including integration footer into device code compilation, depending on that we can update our design document

Reply:
I thought we already decided on a integration footer, right?

Reply:
The built-in is used twice in the codebase:
1. within `kernel_handler::get_specialization_constant` method - *device code*
2. within `handler::set_specialization_constant` method - *host code*

The first one is used by `sycl-post-link` to identify specialization constants and generate corresponding device image properties.
The second one is used by DPC++ RT to connect `specialization_id`s to device image properties.

So, for (1) we call the built-in directly as we are in device code and we can rely on its availability. For (2) we can't do that, because we would like to support 3rd-party host compilers and therefore we rely on the compiler here, in particular on the fact that integration footer will provide to us pre-computed result of calling that built-in without actually putting the call into the source code.

This comes down to a requirement that even though we should get unique strings for internal variables from different translation units, we should get the same string if we call the built-in second time for the same variable within a single translation unit

Reply:
Right, ok, I think we're on the same page.  The value at host-side would be pre-computed by the device side.  We can do the 'same TU, same answer' thing.

Reply:
I thought there were cases where the device compiler is invoked more than once over the same TU.  Doesn't this happen when there are multiple targets specified on "-fsycl-targets"?  In such a case, I presume each invocation of the device compiler would need to return the same string for a call to `__builtin_unique_stable_name(S)` (given the same variable `S`).  Will this be possible when `S` has internal linkage and the returned string is a GUID?

Reply:
@gmlueck THAT part wouldn't be possible, no.

Reply:
Ok, we have a problem here: for a translation unit we have only one host compilation and we have a separate FE invocation for each target.

All of those above should be getting the same strings/IDs assigned to the same variables, BUT at the same time if we have the same variable with internal linkage in separate translation unit, then IDs used there shouldn't interfere with IDs here.

What if we base our unique string generation on some common seed/prefix, which is unique for the translation unit, but common for all FE invocations for that translation unit? I.e. we do generate `GUID`, but we generate it somewhere in driver and pass to all FE invocations. FE in turn, uses mangled name as-is for external variables as an ID and for internal variables it prefixes their mangled names with the prefix passed by driver, which would be common between all target within a single driver invocation, i.e. common for the translation unit?

Reply:
Ah, yeah, we can't really do that with multiple device compilations.  

I don't really have any idea what we could do for a prefix here to auto-gen it, the individual TUs aren't easy to identify, main file names aren't possible since it is possible/somewhat common to usethe same file for multiple TUs.

The idea of having the driver pass some sort of unique token to each is possible, at that point we can just do away with the GUID and do the normal mangling for the internal types.  Though, of course, that can be affected by macros in some cases...

Reply:
> I don't really have any idea what we could do for a prefix here to auto-gen it, the individual TUs aren't easy to identify, main file names aren't possible since it is possible/somewhat common to usethe same file for multiple TUs.

Absolute path + all user-defined macro as an identifier of a TU?

Reply:
Well... maybe? that would be really rough and I'm not sure makes it unique.  There are some constexpr/template tricks to cause 'counting' that might defeat it, but none I'm smart enough to write myself.  

I think I like the idea of having the driver pass us a unique token per-TU during the device compile instead.  That way it can just pass us some random value and have it be the same to each invocation.

Reply:
> The idea of having the driver pass some sort of unique token to each is possible, at that point we can just do away with the GUID and do the normal mangling for the internal types. Though, of course, that can be affected by macros in some cases...

This seems like a workable direction.  The driver creates a GUID and passes it via a command line parameter to the device compiler.  For a specialization constant with internal linkage, the device compiler generates a string "GUID" + "mangled name".

I didn't understand @erichkeane's concern about "that can be affected by macros in some cases...".  Can you elaborate?

Reply:
@gmlueck : A sometimes used pattern (particularly in some of the boost code generators IIRC) is to generate a single .cpp file to represent multiple TUs and execute it multiple times with predefined macros.  So something like:

clang++ My_File.cpp -DALL
clang++ My_File.cpp -DSECTION1
clang++ My_File.cpp -DSECTION2
clang++ My_File.cpp -DSECTION3

etc.  The nasty part about this is sometimes there are code-generation/modification steps in between each, so My_File.cpp isn't actually the same file (let alone having different sections compiled each time!).  So something like:
clang++ My_File.cpp -DALL
clang++ My_File.cpp -DSECTION1
./SomeScriptThatInsertsInfoFromThePreviousThingIntoMyFile
clang++ My_File.cpp -DSECTION2
clang++ My_File.cpp -DSECTION3

Reply:
I see.  I think this is not a problem for the proposed design where the driver creates the GUID.  In the case you describe, each invocation of "clang++" (e.g. "clang++ My_File.cpp -DALL") would generate a different GUID, so each TU will still get unique strings for internal spec constants, despite being compiled from the same source file.

Of course, another solution would be for the driver to have some new pass that created the integration footer before any invocations of the device compiler.  In this case, the footer could be appended to the source file before invoking the device compiler, and then we wouldn't need `__builtin_unique_ID()` at all.  In some ways, this seems like a cleaner design, but I presume we are concerned about extra compilation time overhead by adding a new pass.

Reply:
Let me point out that anything involving full path name is not going to work from a privacy perspective.  That potentially leaks private user identifiable data, such as login id.

The driver creating GUID sounds OK